### PR TITLE
PSGD-Kron's helper functions

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
Building the helper functions for the PSGD Kron optimizer By Xilin-Li. 

Includes subspace-iteration based spectral lower bound calculation, an online procrustes solver and functions to apply the kronecker-factored upper triangular factored preconditioner.

### CODE REVIEW

File: emerging_optimizers/psgd/procrustes_step.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from emerging_optimizers.psgd.psgd_utils import norm_lower_bound_skew
+
+
+__all__ = [
+    "procrustes_step",
+]
+
+
+def procrustes_step(Q: torch.Tensor, max_step_size: float = 1 / 8) -> None:
+    r"""One step of an in-place online solver for the orthogonal Procrustes problem.
```
Comment (skyw) [inline emerging_optimizers/psgd/procrustes_step.py]: Avoid inplace generally. Return the new Q is better.

File: emerging_optimizers/psgd/procrustes_step.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    The orthogonal Procrustes problem is min_U || U Q - I ||_F,   s.t. U^H U = I
```
Comment (skyw) [inline emerging_optimizers/psgd/procrustes_step.py]: missing :math:.

And let's do a follow up PR to add them to auto generated doc.

File: emerging_optimizers/psgd/procrustes_step.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import emerging_optimizers.utils as utils
+from emerging_optimizers.psgd.psgd_utils import norm_lower_bound_skew
+
+
+__all__ = [
+    "procrustes_step",
+]
+
+
+def procrustes_step(Q: torch.Tensor, max_step_size: float = 1 / 8) -> None:
```
Comment (skyw) [inline emerging_optimizers/psgd/procrustes_step.py]: use 0.125 instead of 1/8. Same for other values.

File: emerging_optimizers/psgd/procrustes_step.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    `max_step_size` should be less than 1/4 as we only expand exp(a R) to its 2nd order term.
+
+    This method is a second order Taylor expansion of a Lie algebra parametrized rotation that
+    uses a simple approximate line search to find the optimal step size, from Xi-Lin Li.
+
+    Args:
+        Q: Tensor of shape (n, n), general square matrix to orthogonalize.
+        max_step_size: Maximum step size for the line search. Default is 1/8.
+    """
+    with utils.fp32_matmul_precision("highest"):
```
Comment (skyw) [inline emerging_optimizers/psgd/procrustes_step.py:44]: Add comment why it must use FP32.

File: emerging_optimizers/psgd/procrustes_step.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        R = Q.T - Q
+        R /= norm_lower_bound_skew(R) + torch.finfo(R.dtype).smallest_normal
```
Comment (skyw) [inline emerging_optimizers/psgd/procrustes_step.py]: `torch.finfo(R.dtype).smallest_normal` is too small. use a eps and give it usually 1e-8 or 1e-7.
Use max() is better to avoid dividing by 0.

File: emerging_optimizers/psgd/procrustes_step.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        RQ = R @ Q
+        # trace of RQ is always positive,
+        # since tr(RQ) = ⟨R, Q⟩_F = ⟨Q^T - Q, Q⟩_F = ||Q||_F^2 - ⟨Q, Q⟩_F = ||Q||_F^2 - tr(Q^T Q) ≥ 0
+        tr_RQ = torch.trace(RQ)
+        RRQ = R @ RQ
+        tr_RRQ = torch.trace(RRQ)
+        # clip step size to max_step_size, based on a 2nd order Taylor expansion.
+        step_size = torch.clamp(-tr_RQ / tr_RRQ, min=0, max=max_step_size)
+        # If tr_RRQ >= 0, the quadratic approximation is not concave, we fallback to max_step_size.
+        a = torch.where(tr_RRQ < 0, step_size, max_step_size)
```
Comment (skyw) [inline emerging_optimizers/psgd/procrustes_step.py]: Avoid single letter variable name. This can be name as `xxx_mask`

File: emerging_optimizers/psgd/procrustes_step.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        # rotate Q as exp(a R) Q ~ (I + a R + a^2 R^2/2) Q with an optimal step size by line search
+        # for 2nd order Taylor expansion, only expand exp(a R) to its 2nd term.
+        Q.add_(a * (RQ + 0.5 * a * RRQ))
```
Comment (skyw) [inline emerging_optimizers/psgd/procrustes_step.py]: Return Q instead of inplace. And also use addmm and add etc. when they can save memory traffic.

File: emerging_optimizers/psgd/psgd_kron_contractions.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def partial_contraction(G1: torch.Tensor, G2: torch.Tensor, axis: int) -> torch.Tensor:
+    """Compute the partial contraction of G1 and G2 along axis `axis`.
+    This is the contraction of the two tensors, but with all axes except `axis` contracted.
+
+    Args:
+        G1: Tensor of shape (d_0, d_1, ..., d_{axis-1}, d_{axis}, d_{axis+1}, ..., d_N)
+        G2: Tensor of shape (d_0, d_1, ..., d_{axis-1}, d_{axis}, d_{axis+1}, ..., d_N)
+        axis: int, the axis to contract along
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_kron_contractions.py:35]: Suggestion: add an example, tensordot is not very easy to understand.

File: emerging_optimizers/psgd/psgd_kron_contractions.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    If Q is a vector, we multiply X by Q.
+    If Q is a matrix, we contract Q's second index with X's `axis` index.
+
+    Args:
+        Q_list: List of Q (e.g. the Kronecker factors).
+        X: Tensor of shape (d_0, d_1, ..., d_{axis-1}, d_{axis+1}, ..., d_N)
+    """
+    Q = Q_list[axis]
+    if Q.dim() == 1:
+        shape = [1] * X.dim()
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_kron_contractions.py:147]: Nit: Use squeeze and unsqueeze is more PyTorch.

File: emerging_optimizers/psgd/psgd_kron_contractions.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        shape[axis] = Q.size(0)
+        return X * Q.view(shape)
+    else:
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_kron_contractions.py]: "else" is not necessary, remove.

File: emerging_optimizers/psgd/psgd_kron_contractions.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    Helper function for `_apply_single_kronecker_factor`.
+    If M is (d_out, d_in) we contract M’s second index with X’s `mode` index.
+    `torch.tensordot` is used to contract the two tensors, and then the result is permuted to move the new axis 0 to position `mode`.
+    Returns a new tensor of the same rank, but with size[mode] replaced by d_out.
+    Note that d_{mode} == d_in.
+
+    Args:
+        X: Tensor of shape (d_0, d_1, ..., d_{mode-1}, d_{mode}, d_{mode+1}, ..., d_N)
+        M: Tensor of shape (d_out, d_in)
+        mode: int, the mode to contract with M, with d_{mode} == d_in
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_kron_contractions.py]: Mode doesn't sound like the best name for integer. would `contract_dim` be a right name? or other more informative `xxx_dim`.

File: emerging_optimizers/psgd/psgd_kron_contractions.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+    Returns:
+        Tensor of shape (d_0, d_1, ..., d_{mode-1}, d_out, d_{mode+1}, ..., d_N)
+
+    Example:
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_kron_contractions.py]: Use RST example format.

File: emerging_optimizers/psgd/psgd_utils.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+__all__ = [
+    "balance_q_in_place",
+    "norm_lower_bound_spd",
+    "norm_lower_bound_skew",
+]
+
+
+@torch.compile  # type: ignore[misc]
+def balance_q_in_place(Q_list: List[torch.Tensor]) -> None:
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_utils.py]: Balance is not very informative. consider informative names like  `xxx_norm` or `xxx_scale`

File: emerging_optimizers/psgd/psgd_utils.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    # Compute max-abs norm of each factor
+    norms = [torch.max(torch.abs(Q)) for Q in Q_list]
+
+    # Compute geometric mean of those norms
+    stacked = torch.stack(norms)
+    gmean = torch.prod(stacked) ** (1.0 / order)
+
+    # Rescale each factor so its max‐abs entry == geometric mean
+    for Q, norm in zip(Q_list, norms, strict=True):
+        Q.mul_(gmean / norm)
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_utils.py:69]: Future improvement: I lot of code in this PR can/should be optimized. Sequence of operation is very inefficient. Although coming up with a better performing counterpart is some time non-trivial.

File: emerging_optimizers/psgd/psgd_utils.py [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+
+@torch.compile  # type: ignore[misc]
+def norm_lower_bound_spd(A: torch.Tensor, k: int = 4, half_iters: int = 2) -> torch.Tensor:
+    r"""Returns a cheap lower bound for the spectral norm of a symmetric positive definite matrix.
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_utils.py]: 3rd party singular is not necessary in one liner.

File: emerging_optimizers/psgd/psgd_utils.py [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+        A: Tensor of shape :math:`(n, n)`, symmetric positive definite.
+        k: Dimension of the subspace.
+        half_iters: Half of the number of subspace iterations.
+    Returns:
+        A scalar giving a lower bound on :math:`\\|A\\|_2`.
+    """
+
+    # Compute normalizing factor from the largest diagonal entry to prevent overflow/underflow and use smallest representable normal number for numerical stability
+    normalization = A.diagonal().amax() + torch.finfo(A.dtype).smallest_normal
+    A = A / normalization
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_utils.py]: use
```python
A = A / max(A.diagonal().amax(), eps)
```

Don't use `torch.finfo(A.dtype).smallest_normal` as eps, same for other places in the PR.

File: tests/test_procrustes_step.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+from emerging_optimizers.psgd.procrustes_step import procrustes_step
+from emerging_optimizers.utils import fp32_matmul_precision
+
+
+class ProcrustesStepTest(parameterized.TestCase):
+    """Test cases for procrustes_step function."""
+
+    def setUp(self) -> None:
+        """Set up test fixtures."""
+        torch.manual_seed(42)
```
Comment (skyw) [inline tests/test_procrustes_step.py]: Don't reset seed per test. can do it globally for repeatability.

File: tests/test_procrustes_step.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+        self.device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
```
Comment (skyw) [inline tests/test_procrustes_step.py]: Can use flags, don't auto select device.

File: tests/test_procrustes_step.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+    def test_improves_orthogonality_simple_case(self) -> None:
+        """Test that procrustes_step doesn't worsen orthogonality for a simple case."""
+
+        # Make a SPD non-orthogonal matrix
+        Q = torch.randn(2, 2, device=self.device, dtype=torch.float32)
+        Q = Q @ Q.T
+
+        initial_obj = self._procrustes_objective(Q)
+
+        procrustes_step(Q, max_step_size=1 / 16)
```
Comment (skyw) [inline tests/test_procrustes_step.py:51]: write 0.0625 instead of 1/16.

File: tests/test_procrustes_step.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+        self.assertLessEqual(final_obj.item(), initial_obj.item() + 1e-6)
+
+    def test_modifies_matrix_in_place(self) -> None:
+        """Test that procrustes_step modifies the matrix in place."""
+        Q = torch.randn(3, 3, device=self.device)
+        Q_original_id = id(Q)
+
+        procrustes_step(Q, max_step_size=1 / 16)
+
+        self.assertEqual(id(Q), Q_original_id)
```
Comment (skyw) [inline tests/test_procrustes_step.py]: Favor Q is Q_original instead of comparing id.

File: tests/test_procrustes_step.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+        self.assertLess(final_obj.item(), 1e-6)
+        self.assertLess(final_obj.item(), initial_obj.item() + 1e-6)
+
+    @parameterized.parameters(
+        (1 / 64,),
+        (1 / 32,),
+        (1 / 16,),
+        (1 / 8,),
+    )
+    def test_different_step_sizes(self, max_step_size: float) -> None:
```
Comment (skyw) [inline tests/test_procrustes_step.py]: Suggestion: include designated behavior in test name. e.g. test_xxx_what_smaller_than_what.
Same for all other test names.

File: tests/test_procrustes_step.py [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+        (512,),
+        (8192,),
+    )
+    def test_different_matrix_sizes(self, size: int) -> None:
+        """Test procrustes_step improvement with different matrix sizes."""
+        # Create a non-orthogonal matrix by scaling an orthogonal one
+        A = torch.randn(size, size, device=self.device, dtype=torch.float32)
+        with fp32_matmul_precision("highest"):
+            Q_orth, _ = torch.linalg.qr(A)
+        Q = Q_orth + 1e-2 * torch.randn(size, size, device=self.device, dtype=torch.float32) / math.sqrt(size)
```
Comment (skyw) [inline tests/test_procrustes_step.py:129]: comment significance of magic numbers.

File: tests/test_psgd_contractions.py [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+        """Set up test fixtures."""
+        torch.manual_seed(42)
+        self.device = torch.device("cuda")
+
+    @parameterized.parameters(
+        (2, 3, 3),
+        (2, 3, 4),
+        (2, 3, 5),
+    )
+    def test_partial_contraction_matches_reconstructed(self, size1: int, size2: int, size3: int) -> None:
```
Comment (skyw) [inline tests/test_psgd_contractions.py:46]: This is a better test name than many other. 👍

File: emerging_optimizers/psgd/psgd_utils.py [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+
+    Args:
+        Q_list: List of Q (e.g. the Kronecker factors), each tensor will be modified in place.
+
+    Returns:
+        None
+
+    """
+    if (order := len(Q_list)) <= 1:
+        return
```
Comment (skyw) [inline emerging_optimizers/psgd/psgd_utils.py:59]: This should be a TypeError. Launching a function with empty list is generally considered wrong usage.

Also use `if list_name` to determine whether the list is empty.

File: tests/test_procrustes_step.py (hunk 20 above)
Comment (mkhona-nvidia) [reply tests/test_procrustes_step.py]: the value of Q changes, so this will be false

File: emerging_optimizers/psgd/psgd_utils.py (hunk 13 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/psgd/psgd_utils.py]: replaced with "uniformize"

File: emerging_optimizers/psgd/psgd_kron_contractions.py (hunk 8 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/psgd/psgd_kron_contractions.py:35]: We have tensordot documentation in the docs

File: tests/test_procrustes_step.py (hunk 20 above)
Comment (mkhona-nvidia) [reply tests/test_procrustes_step.py]: actually, with removing in-place, this test is unnuecssary

File: emerging_optimizers/psgd/psgd_utils.py (hunk 16 above)
Comment (skyw) [reply emerging_optimizers/psgd/psgd_utils.py]: normalization  is a bad name for temporary variable here, meaning is not quite right.
we should favor max(something, eps) unless there is a convention being established. Since torch normalize use this, just write it as I suggested above. Same for other similar ones.
